README.md
# invoice_api

## What does it do?
This is an API which helps you to create European invoices.
- This API can generate check digits with a given reference (invoice number).
- It can validate a refence number by using the checkdigits.
- It can store, retrieve, update and delete invoice data.

## Getting Started
### Install

1. open the terminal at the api folder
2. run `npm install`

### Run docker container

1. open the terminal
2. run `docker-compose up`

## API endpoints

### Request

`GET /get-check-digits/:reference`
    
    http://localhost:3000/get-check-digits/reference

| Parameter | Type | Description | Example |
| :--- | :--- | :--- | :--- |
| `reference` | `integer` | **Required**. An invoice reference | 5496842634654

### Response

    check digits
    example: "06"



### Request

`GET /validate/:referenceNumber/:checkDigits`
    
    http://localhost:3000/referenceNumber/checkDigits

| Parameters | Type | Description | Example |
| :--- | :--- | :--- | :--- |
| `referenceNumber` | `string` | **Required**. An invoice reference number | RF53810888428933129078686
| `checkDigits` | `integer` | **Required**. check digits of the refrence number | 53

### Response

    returns true if refrence number and check digits are valid.
    if invalid it returns: "invalid reference number."





### Request

`POST /create-invoice`
    
    http://localhost:3000/create-invoice

| Parameters | Type | Description | Example |
| :--- | :--- | :--- | :--- |
| `reference_number` | `string` | **Required**. A invoice reference number | RF0812318152
| `buisiness_name` | `string` | **Required**. Your business name | My business
| `client_name` | `string` | **Required**. Your client name | Google
| `amount_total` | `number` | **Required**. Total payment amunt | 1001.02
| `invoice_number` | `integer` | **Required**. The uniqe number of your invoice | 1002
| `due_date` | `date` | **Required**. The due date of the invoice format: yyyy-mm-dd | 2021-08-01
| `type_id` | `integer` | **Required**. The type id of your invoice see type id's in availeble invoice types. | 2

### Response

    returns status code 201 and message 'invoice created succesfully.' if request was send succesfully.
    if invalid it returns status code 400 and a message explaining whats wrong.    



### Request

`GET /get-by-invoice-number/:invoiceNumber`
    
    http://localhost:3000/get-by-invoice-number/invoiceNumber

| Parameters | Type | Description | Example |
| :--- | :--- | :--- | :--- |
| `invoiceNumber` | `integer` | **Required**. The uniqe number of your invoice | 1002

### Response

    returns the data of the invoice in json format
    example response:

```javascript
{
    "invoice": {
        "id": 1,
        "uuid": "2acb89a0-4f67-11eb-8b3e-0b8e106cdd95",
        "reference_number": "RF0812318152",
        "business_name": "My business",
        "client_name": "google",
        "amount_total": "10000",
        "invoice_number": "1002",
        "due_date": "2000-01-02",
        "type_id": 1,
        "created_at": "2021-01-05T15:03:28.315Z",
        "updated_at": "2021-01-05T15:03:28.315Z"
    }
}
```

### Request

`PUT /update-invoice`
    
    http://localhost:3000/update-invoice/invoiceNumber

| Parameters | Type | Description | Example |
| :--- | :--- | :--- | :--- |
| `invoiceNumber` | `integer` | **Required**. The uniqe number of your invoice **(URL param)** | 1002
| `referenceNumber` | `string` | **Required**. A invoice reference number **(Body param)** | RF0812318152
| `buisiness_name` | `string` | **Required**. Your business name **(Body param)** | My business
| `client_name` | `string` | **Required**. Your client name **(Body param)** | Google
| `amount_total` | `number` | **Required**. Total payment amunt **(Body param)** | 1001.02
| `invoice_number` | `integer` | **Required**. The uniqe number of your invoice **(Body param)** | 1002
| `due_date` | `date` | **Required**. The due date of the invoice format: yyyy-mm-dd **(Body param)** | 2021-08-01
| `type_id` | `integer` | **Required**. The type id of your invoice see type id's in availeble invoice types. **(Body param)** | 2

### Response

    returns status code 200 and message 'invoice updated succesfully.' if request was send succesfully.
    if invalid it returns status code 400 and a message explaining whats wrong.  


### Request

`DELETE /delete-invoice/:invoiceNumber`
    
    http://localhost:3000/delete-invoice/invoiceNumber

| Parameters | Type | Description | Example |
| :--- | :--- | :--- | :--- |
| `invoiceNumber` | `integer` | **Required**. The uniqe number of the invoice you want to delete. | 1002

### Response

    returns status code 200 and message 'invoice deleted succesfully.' if request was send succesfully.
    if invalid it returns status code 400 and a message explaining whats wrong.  


## Availeble invoice types

| id | name |
| :--- | :--- |
| `Standard invoice` | `1`
| `Credit invoice` | `2`
| `Expence report` | `3`
| `Debit invoice` | `4`
| `Mixed invoice` | `5`
| `Commercial invoice` | `6`
| `Timesheet invoice` | `7`
| `Pro forma invoice` | `8`
| `Intrim invoice` | `9`
| `Final invoice` | `10`
| `Past due invoice` | `11`
| `Recurring invoice` | `12`
| `E-invoice` | `13`

## Where can I find help?
If you have a question you can email me at *seppe.clottemans@student.ehb.be*.
## Project Status
This project is currently **in development**.
## Authors
* Seppe Clottemans


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1.0] - 2021-01-15

### Added

- End to end test of the API.

### Fixed

- Mistakes in endpoints documentation fixed.
- Compares json response after delete invoice instead of text response. 

## [2.0.0] - 2021-01-15

### Added

- Fully working api endpoint to store an invoice in the database.
- Fully working api endpoint to get an invoice from the database by its invoice number.
- Fully working api endpoint to update an invoice in the database.
- Fully working api endpoint to delete an invoice in the database.
- Integration tests for the: create-invoice, get-by-invoice-number, update-invoice and delete-invoice endpoints. 
- Documentation for each of these enpoints.

## [1.4.0] - 2021-01-15

### Added

- Delete-invoice api endpoint.
- Integration tests for the delete-invoice api endpoint.
- Integration test for trying to update an unexisting invoice.
- Documentation for the delete-invoice api endpoint.

## [1.3.1] - 2021-01-15

### Fixed

- Succesfully creating an invoice returns status code 201 instead of 202.

## [1.3.0] - 2021-01-15

### Added

- update-invoice api endpoint.

### Fixed

- amount_total changed in database from a bigInteger to a double.
- Retrieve dates in correct format from database.
- Integration tests for the update-invoice api endpoint.

## [1.2.3] - 2021-01-14

### Added

- Integration tests for the update-invoice api endpoint.

## [1.2.2] - 2021-01-12

### Added

- Integration tests for the create-invoice api endpoint.
- Integration tests for the get-by-invoice-number api endpoint.

### Fixed

- ci pipeline doesn't run integration tests.

## [1.2.1] - 2021-01-12

### Added

- documentation for each function added containing: a description of what the function or endpoint does, all needed parameters and expected types and what the function or endpoint returns.
- missing indentations.

### Fixed

- fixed mistake in readme file.

## [1.2.0] - 2021-01-05

### Added

- Fully working api endpoint to get invoices by invoice number.
- Documentation for the create and get invoice endpoints.

## [1.1.1] - 2021-01-01

### Added

- Check if invoice doesn't already exists before creation.
- Check if invoice type id exists. 

## [1.1.0] - 2020-12-30

### Added

- Fully working api endpoint to store an invoince in the database.
- Parameter validation functions for the create invoice api endpoint.
- Unit tests for the Parameter validation functions in the databaseHelper file.
- Database structure on project build.
- invoice_types table automaticlly filled on build.

## [1.0.0] - 2020-12-30

### Description

- Fully working api endpoint to validate a reference number.
- Fully working api endpoint to get check digits from a reference.
- Unit tests for each of these helper functions. 

## [0.1.2] - 2020-12-30

### Added

- Fully working api endpoint to validate a reference number.
- Unit tests for the Helper function to validate a reference number.

## [0.1.1] - 2020-12-29

### Fixed

- Fixed a bug that when the number was too big the check digits would be incorrect.

## [0.1.0] - 2020-12-29

### Added

- Fully working api endpoint to get check digits from a reference.

## [0.0.2] - 2020-12-29

### Added

- Unit tests for the Helper function to get check digits from a reference.

## [0.0.1] - 2020-12-28

### Added

- API endpoint to get check digits from a reference.
- Helper function to get check digits from a reference.


api/src/_tests_/helpers.test.js
const Helpers = require('./../utils/helpers');

describe('check if uuid is valid', () => {
    
    test('string must be a valid uuid', () => {
        expect(Helpers.checkValidUuid('This is not a valid uuid')).toBeFalsy();
    });

    test('valid uuid should return true', () => {
        expect(Helpers.checkValidUuid('ef73e7b0-2fec-11eb-988f-77262bbf5d7e')).toBe(true);
        expect(Helpers.checkValidUuid('31c30b40-2fee-11eb-b22a-ebac5daea745')).toBe(true);
    });

})

api/src/__tests__/helpers.test.js
const Helpers = require('../utils/helpers');

describe('check if uuid is valid', () => {
    
    test('string must be a valid uuid', () => {
        expect(Helpers.checkValidUuid('This is not a valid uuid')).toBeFalsy();
    });

    test('valid uuid should return true', () => {
        expect(Helpers.checkValidUuid('ef73e7b0-2fec-11eb-988f-77262bbf5d7e')).toBe(true);
        expect(Helpers.checkValidUuid('31c30b40-2fee-11eb-b22a-ebac5daea745')).toBe(true);
    });

});

// get check digits tests

describe('check if given reference returns valid check digits', () => {
    
    //examples tested on https://www.mobilefish.com/services/creditor_reference/creditor_reference.php
    references = ['5496842634654', '12318152 ', 454, '36524564', 78975, '9875614324', '810888428933129078686'];
    expectedCheckDigits = ['06', '08', '65', '80', '30', '42', '53'];

    let i = 0;
    test.each(references)(
        'valid references should return the expected check digits',
        (reference) => {
            expect(Helpers.getCheckDigits(reference)).toStrictEqual([200, expectedCheckDigits[i]]);
            i++;
        },
    );

});

describe('check if given invalid reference returns error status with error message', () => {
    
    // test with letters
    test('invalid reference strings returns error status with error message', () => {
        expect(Helpers.getCheckDigits('4203B224A')).toStrictEqual([400, "The reference needs to be an integer. It can not contain any letters."]);
        expect(Helpers.getCheckDigits('RF47374')).toStrictEqual([400, "The reference needs to be an integer. It can not contain any letters."]);
        expect(Helpers.getCheckDigits('febfjezbf')).toStrictEqual([400, "The reference needs to be an integer. It can not contain any letters."]);
    });

    // test with invalid types
    test('invalid reference types returns error status with error message', () => {
        expect(Helpers.getCheckDigits({})).toStrictEqual([400, "The reference needs to be an integer. It can not contain any letters."]);
        expect(Helpers.getCheckDigits({"test": 12345})).toStrictEqual([400, "The reference needs to be an integer. It can not contain any letters."]);
        expect(Helpers.getCheckDigits([])).toStrictEqual([400, "The reference needs to be an integer. It can not contain any letters."]);
    });

    // test with negative numbers
    test('negative references returns error status with message', () => {
        expect(Helpers.getCheckDigits(-454)).toStrictEqual([400, "The reference can not be negative."]);
        expect(Helpers.getCheckDigits(-12318152)).toStrictEqual([400, "The reference can not be negative."]);
        expect(Helpers.getCheckDigits('-12318152')).toStrictEqual([400, "The reference can not be negative."]);
        expect(Helpers.getCheckDigits('-5496842634654')).toStrictEqual([400, "The reference can not be negative."]);
    });

});


// validate reference number tests

describe('check if given valid credit reference pass validation check', () => {
    
    //examples tested on https://www.mobilefish.com/services/creditor_reference/creditor_reference.php
    creditorReferences = ['RF065496842634654', 'RF0812318152', 'RF65454', 'RF8036524564', 'RF3078975', 'RF429875614324'];
    checkDigits = ['06', '08', '65', '80', '30', '42'];

    let i = 0;
    test.each(creditorReferences)(
        'validation should return true',
        (referenceNumber) => {
            expect(Helpers.validateReferenceNumber(referenceNumber, checkDigits[i])).toBe(true);
            i++;
        },
      );

});

describe('check if given invalid credit reference fails the validation check', () => {
    
    //examples tested on https://www.mobilefish.com/services/creditor_reference/creditor_reference.php
    creditorReferences = ['RF065496842634653', 'RF0812318151', 'RF65458', 'RF8036524569', 'RF3078974', 'RF429875614322', 'RF429875614322'];
    checkDigits = ['06', '08', '65', '80', '30', '42', '42'];

    let i = 0;
    test.each(creditorReferences)(
        'validation should return false',
        (referenceNumber) => {
            expect(Helpers.validateReferenceNumber(referenceNumber, checkDigits[i])).toBeFalsy();
            i++;
        },
      );

});

describe('check if given invalid credit reference fails the validation check', () => {
    
    test('invalid parameters return false on validation', () => {
        expect(Helpers.validateReferenceNumber('58', 10)).toBeFalsy();
        expect(Helpers.validateReferenceNumber('', 1)).toBeFalsy();
        expect(Helpers.validateReferenceNumber({}, '10')).toBeFalsy();
        expect(Helpers.validateReferenceNumber([], '10')).toBeFalsy();
    });

});

api/src/__tests__/end_to_end.test.js
const supertest = require('supertest');
const http = require('http');

const app = require('../server.js');
const request = supertest(app);

// randomly select an invoice number.
const invoiceNumber = Math.floor(Math.random() * 100000);

let invoice = {
    reference_number: "",
    business_name: "seppe corp",
    client_name: "apple",
    amount_total: 1024851,
    invoice_number: invoiceNumber,
    due_date: "2023-08-15",
    type_id: 8
}

const reference = 6017710223;

let checkDigits = "";
let referenceNumber = "";

describe('get check digits for the given reference and validate the invoice reference number.', () => {

    test('get check digits for the given reference', async (done) => {

        const getCheckDigitsResponce = await request.get(`/get-check-digits/${reference}`);

        expect(getCheckDigitsResponce.status).toStrictEqual(200);

        // return type should be string
        expect(typeof getCheckDigitsResponce.text).toStrictEqual("string");

        // always should return 2 numbers
        expect(getCheckDigitsResponce.text.length).toStrictEqual(2);

        checkDigits = getCheckDigitsResponce.text;

        // reference number consists of RF, the checkdigits and the reference.
        referenceNumber = "RF" + getCheckDigitsResponce.text + reference

        done();

    });

    test('validate if the given reference number is valid with the given checkdigits', async (done) => {

        const validateReferenceNumberResponce = await request.get(`/validate/${referenceNumber}/${checkDigits}`);

        expect(validateReferenceNumberResponce.status).toStrictEqual(200);

        // if valid referenceNumber it should return true.
        expect(typeof validateReferenceNumberResponce.body).toStrictEqual("boolean");
        expect(validateReferenceNumberResponce.body).toStrictEqual(true);

        done();

    });

});

// create/store a new invoice in the database and retrieve the newly created invoice.
describe('Create a new invoice in the database and check if it stored', () => {

    test('delete invoice if there already was one in the database then create a new invoice.', async (done) => {

        // get invoice
        const getInvoiceResponse = await request.get(`/get-by-invoice-number/${invoiceNumber}`);

        expect(getInvoiceResponse.status).toStrictEqual(200);
        
        // check if invoice already exists
        if(!Object.keys(getInvoiceResponse.body).length == 0 && getInvoiceResponse.body.constructor == Object){
            // an invoice with this number already exists

            // delete existing invoice
            const deleteInvoiceResponce = await request.delete(`/delete-invoice/${invoiceNumber}`);

            expect(deleteInvoiceResponce.status).toStrictEqual(200);
            expect(deleteInvoiceResponce.text).toStrictEqual('invoice deleted succesfully.');
        }

        // create new invoice
        invoice.invoice_number = invoiceNumber;
        const createInvoiceResponce = await request.post('/create-invoice').send(invoice);

        expect(createInvoiceResponce.text).toStrictEqual("invoice created succesfully.");
        expect(createInvoiceResponce.status).toStrictEqual(201);



        done();
    });

    test('check if invoice was actually stored in the database.', async (done) => {

        // get invoice
        const getInvoiceResponse = await request.get(`/get-by-invoice-number/${invoiceNumber}`);

        expect(getInvoiceResponse.status).toStrictEqual(200);

        // check if the created invoice has all values
        expect(getInvoiceResponse.body.invoice.reference_number).toStrictEqual(invoice.reference_number);
        expect(getInvoiceResponse.body.invoice.business_name).toStrictEqual(invoice.business_name);
        expect(getInvoiceResponse.body.invoice.client_name).toStrictEqual(invoice.client_name);
        expect(getInvoiceResponse.body.invoice.amount_total).toStrictEqual(invoice.amount_total);
        expect(getInvoiceResponse.body.invoice.due_date).toStrictEqual(invoice.due_date);
        expect(getInvoiceResponse.body.invoice.type_id).toStrictEqual(invoice.type_id);
        expect(parseInt(getInvoiceResponse.body.invoice.invoice_number)).toStrictEqual(invoice.invoice_number);


        done();
    });

});


// update the newly created invoice and check if the values acctually changed.
describe('Update invoice and check if the values acctually changed.', () => {

    test('update invoice with new values', async (done) => {

        invoice.reference_number = "RF65454"
        invoice.business_name = "new seppe corp"
        invoice.client_name = "facebook"
        invoice.amount_total = 16010120.61
        invoice.due_date = "2024-12-23"
        invoice.type_id = 10

        const updateInvoiceResponce = await request.put(`/update-invoice/${invoiceNumber}`).send(invoice);

        expect(updateInvoiceResponce.text).toStrictEqual("invoice updated succesfully.");
        expect(updateInvoiceResponce.status).toStrictEqual(200);

        done();
    });

    test('check if invoice was actually updated in the database.', async (done) => {

        // get invoice
        const getInvoiceResponse = await request.get(`/get-by-invoice-number/${invoiceNumber}`);

        expect(getInvoiceResponse.status).toStrictEqual(200);

        // check if the created invoice has all values
        expect(getInvoiceResponse.body.invoice.reference_number).toStrictEqual(invoice.reference_number);
        expect(getInvoiceResponse.body.invoice.business_name).toStrictEqual(invoice.business_name);
        expect(getInvoiceResponse.body.invoice.client_name).toStrictEqual(invoice.client_name);
        expect(getInvoiceResponse.body.invoice.amount_total).toStrictEqual(invoice.amount_total);
        expect(getInvoiceResponse.body.invoice.due_date).toStrictEqual(invoice.due_date);
        expect(getInvoiceResponse.body.invoice.type_id).toStrictEqual(invoice.type_id);
        expect(parseInt(getInvoiceResponse.body.invoice.invoice_number)).toStrictEqual(invoice.invoice_number);


        done();
    });

});


// delete the newly created and updated invoice and check if the invoice does not exist in the database anymore.
describe('Delete invoice and check if the invoice does not exist in the database anymore.', () => {

    test('delete the invoice from the database', async (done) => {

        // delete the invoice
        const deleteInvoiceResponce = await request.delete(`/delete-invoice/${invoiceNumber}`);

        expect(deleteInvoiceResponce.status).toStrictEqual(200);
        expect(deleteInvoiceResponce.text).toStrictEqual("invoice deleted succesfully.");

        done();
    });

    test('check if invoice was actually updated in the database.', async (done) => {

        // get invoice
        const getInvoiceResponse = await request.get(`/get-by-invoice-number/${invoiceNumber}`);

        // check if response is empty
        expect(getInvoiceResponse.status).toStrictEqual(200);
        expect(getInvoiceResponse.body).toStrictEqual({});
        expect(Object.keys(getInvoiceResponse.body).length).toStrictEqual(0);
        expect(getInvoiceResponse.body.constructor).toStrictEqual(Object);


        done();
    });

});

api/src/utils/helpers.js
const {v1: uuidv1 } = require('uuid');
const { validate : uuidValidate }  = require('uuid');

const Helpers = {
  /**
   * Generates a UUID
   * @returns {integer} uuid
   */
  generateUUID: () => {
    const uuid = uuidv1();
    return uuid;
  },

  /**
   * Validates a UUID
   * @param {integer} uuid
   * @returns {boolean} valid or invalid uuid
   */
  checkValidUuid: (uuid) => {
    return uuidValidate(uuid)
  },

  /**
   * Get check digits of the given reference
   * @param {integer} invoiceNumber
   * @returns {Array} array contains status code and a string with the reference number or feedback of the invalid parameter.
   */
  getCheckDigits(invoiceNumber) {
    if (isNaN(invoiceNumber) || typeof invoiceNumber === "object") {
      return [400, "The reference needs to be an integer. It can not contain any letters."];
    }

    if (invoiceNumber < 0) {
      return [400, "The reference can not be negative."];
    }

    // Apply the modulo 97 to the invoice number / reference.
    invoiceNumber = BigInt(invoiceNumber) % BigInt(97);
    //The european ref number is prefixed by RF as an identifier.
    //R = 27
    //F = 15
    //Next add the translated character digits to the end of the reference followed by '00'.
    //Apply modulo 97 again and subtract the result from 98
    let checkDigits = BigInt(98) - (BigInt(invoiceNumber + "271500") % BigInt(97));
    //This result gives you the check digits.
    if (checkDigits == 0) {
      //We always need check digits so if they are 0 we use '00' as check digits.
      return [200, "00"];
    } else if (checkDigits < 10) {
      //We always need 2 digits so if the check digits are smaller then 10 we prepend a '0'.
      return [200, "0" + checkDigits];
    } else {
      return [200, checkDigits.toString()];
    }
  },

  /**
   * Validates if the given referenceNumber is valid by using the checkdigits
   * @param {string} referenceNumber
   * @param {string} checkDigits
   * @returns {boolean} true if valid referenceNumber else false.
   */
  validateReferenceNumber(referenceNumber, checkDigits) {

    if (typeof referenceNumber === "object") {
      return false;
    }

    //Remove all spaces from the reference.
    referenceNumber = referenceNumber.replace(/\s+/g, '');
    //Remove the RF and the check digits.
    referenceNumber = referenceNumber.substring(4);
    //Add them at the end in digit form.
    referenceNumber = referenceNumber + "2715" + checkDigits;
    //Check if modulo 97 returns 1.
    return BigInt(referenceNumber) % BigInt(97) === BigInt(1);
  }

}

module.exports = Helpers

api/src/utils/databaseHelper.js
// all expectedParameters with expected types.
const expectedParameters = [
    {name: 'reference_number', type: 'string'},
    {name: 'business_name', type: 'string'},
    {name: 'client_name', type: 'string'},
    {name: 'amount_total', type: 'number'},
    {name: 'invoice_number', type: 'number'},
    {name: 'due_date', type: 'date'},
    {name: 'type_id', type: 'number'}
];


const databaseHelpers = {

    /**
     * checks if all parameters for creating / updating an invoice are given.
     * @param {object} parameters
     * @returns {Array} index 0 of array is true if all necessary parameters are given else false + string with missing parameter.
     */
    checkInvoiceParameters: (parameters) => {

        // check if all required parameters are given.
        for (let i = 0; i < expectedParameters.length; i++) {
            if (!parameters.hasOwnProperty(expectedParameters[i].name)) {
                return [false, `missing parameter ${expectedParameters[i].name}`]
            }
        }

        return [true]
    },

    /**
     * checks if all parameters for creating / updating an invoice have the right types
     * @param {object} parameters
     * @returns {Array} index 0 of array is true if all parameters have the right type else false + string with invalid parameter.
     */
    checkInvoiceParametertypes: (parameters) => {
        for (const parameter in parameters) {
            const expectedParameter = expectedParameters.find(expectedParameter => {
                return expectedParameter.name == parameter
            });

            // check if parameters have the right type
            if (expectedParameter.type == 'number' && isNaN(parameters[parameter])) {
                return [false, `parameter ${parameter} must be numeric`]
            } else if (expectedParameter.type == 'date' && !databaseHelpers.isValidDate(parameters[parameter])) {
                return [false, `parameter ${parameter} must be a valid date`]
            }
        }

        return [true]
    },

    /**
     * checks if all parameters for creating / updating an invoice have the right types
     * @param {string} dateString
     * @returns {boolean} true if datestring is valid date else false.
     */
    isValidDate: (dateString) => {
        // source: https://stackoverflow.com/questions/18758772/how-do-i-validate-a-date-in-this-format-yyyy-mm-dd-using-jquery
        var regEx = /^\d{4}-\d{2}-\d{2}$/;
        if (!dateString.match(regEx)) return false; // Invalid format
        var d = new Date(dateString);
        var dNum = d.getTime();
        if (!dNum && dNum !== 0) return false; // NaN value, Invalid date
        return d.toISOString().slice(0, 10) === dateString;
    }

}

module.exports = databaseHelpers

api/src/server.js
const express = require('express')
const bodyParser = require('body-parser');
const http = require('http');
const Helpers = require('./utils/helpers.js');
const databaseHelpers = require('./utils/databaseHelper.js');

const types = require('pg').types;
// override parsing date column to Date()
types.setTypeParser(1082, val => val);


const pg = require('knex')({
  client: 'pg',
  version: '9.6',
  searchPath: ['knex', 'public'],
  connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/invoice_api_db'
});


const app = express();
http.Server(app);


app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    // to support URL-encoded bodies
    extended: true
  })
);

app.get('/test', (req, res) => {
  res.status(200).send();
})

/**
 * Returns the check digits of the given reference.
 * @param {integer} reference
 * @returns {string} checkdigits of the given reference
 */
app.get('/get-check-digits/:reference', (req, res) => {
  let result = Helpers.getCheckDigits(req.params.reference);
  res.status(result[0]);
  res.send(result[1]);
})

/**
 * Validates the reference number with the given check digits.
 * @param {string} referenceNumber (invoice number)
 * @param {integer} checkDigits
 * @returns {(boolean|string)} true if valid referenceNumber else string: "invalid reference number."
 */
app.get('/validate/:referenceNumber/:checkDigits', (req, res) => {
  const result = Helpers.validateReferenceNumber(req.params.referenceNumber, req.params.checkDigits);
  if (result) {
    res.send(result);
  } else {
    res.send("invalid reference number.");
  }
})


// database endpoints



/**
 * Creates a new invoice.
 * @param {string} referenceNumber (invoice number)
 * @param {string} buisiness_name
 * @param {string} client_name
 * @param {number} amount_total
 * @param {integer} invoice_number
 * @param {date} due_date
 * @param {integer} type_id
 * @returns {string} returns feedback if invalid requests else returns: "invoice created succesfully."
 */
app.post('/create-invoice', async (req, res) => {
  // check if all parameters are given.
  const parameterGivenCheck = databaseHelpers.checkInvoiceParameters(req.body);

  if (parameterGivenCheck[0]) {
    // check if parameters have the right types
    const parametersTypeCheck = databaseHelpers.checkInvoiceParametertypes(req.body)
    if (parametersTypeCheck[0]) {

      //check if invoice doesn't already exist
      const invoice = await pg.count('invoice_number').from('invoices').where({
        invoice_number: req.body.invoice_number
      })
      if (invoice[0].count == 0) {

        // check if invoice type exists
        const invoiceType = await pg.count('id').from('invoice_types').where({
          id: req.body.type_id
        })
        if (invoiceType[0].count == 1) {
          const uuid = Helpers.generateUUID();
          await pg
            .table('invoices')
            .insert({
              uuid,
              reference_number: req.body.reference_number,
              business_name: req.body.business_name,
              client_name: req.body.client_name,
              amount_total: req.body.amount_total,
              invoice_number: req.body.invoice_number,
              due_date: req.body.due_date,
              type_id: req.body.type_id
            })
          res.status(201).send('invoice created succesfully.');
        } else {
          res.status(400).send('invalid type_id.');
        }

      } else {
        res.status(400).send('invoice with this invoice number already exists');
      }

    } else {
      res.status(400).send(parametersTypeCheck[1]);
    }

  } else {
    res.status(400).send(parameterGivenCheck[1]);
  }
})

/**
 * Gets an invoice by its invoice number.
 * @param {integer} invoiceNumber
 * @returns {(json|string)} json object with the invoice in it. or string with invalid request feedback.
 */
app.get('/get-by-invoice-number/:invoiceNumber', async (req, res) => {
  if (!isNaN(req.params.invoiceNumber)) {
    const invoice = await pg
      .first('*')
      .from('invoices')
      .where({
        invoice_number: req.params.invoiceNumber
      })
    res.json({
      invoice: invoice
    })
  } else {
    res.status(400)
    res.send('invoiceNumber parameter needs to be a number')
  }
})



/**
 * Updates an existing invoice with the new values invoice.
 * Prameters in the URL:
 * @param {integer} invoice_number
 * Body parameters:
 * @param {string} referenceNumber (invoice number)
 * @param {string} buisiness_name
 * @param {string} client_name
 * @param {number} amount_total
 * @param {integer} invoice_number
 * @param {date} due_date
 * @param {integer} type_id
 * @returns {string} returns feedback if invalid requests else returns: "invoice updated succesfully."
 */
app.put('/update-invoice/:invoiceNumber', async (req, res) => {

  if (!isNaN(req.params.invoiceNumber)) {

    // check if all parameters are given.
    const parameterGivenCheck = databaseHelpers.checkInvoiceParameters(req.body);

    if (parameterGivenCheck[0]) {
      // check if parameters have the right types
      const parametersTypeCheck = databaseHelpers.checkInvoiceParametertypes(req.body)
      if (parametersTypeCheck[0]) {

        //check if invoice exists
        const invoice = await pg.count('invoice_number').from('invoices').where({
          invoice_number: req.params.invoiceNumber
        })
        if (invoice[0].count == 1) {

          // check if invoice type exists
          const invoiceType = await pg.count('id').from('invoice_types').where({
            id: req.body.type_id
          })
          if (invoiceType[0].count == 1) {
            await pg
              .table('invoices')
              .where('invoice_number', req.params.invoiceNumber)
              .update({
                reference_number: req.body.reference_number,
                business_name: req.body.business_name,
                client_name: req.body.client_name,
                amount_total: req.body.amount_total,
                due_date: req.body.due_date,
                type_id: req.body.type_id
              })
            res.status(200).send('invoice updated succesfully.');
          } else {
            res.status(400).send('invalid type_id.');
          }

        } else {
          res.status(400).send('An invoice with this invoice number does not exist.');
        }

      } else {
        res.status(400).send(parametersTypeCheck[1]);
      }

    } else {
      res.status(400).send(parameterGivenCheck[1]);
    }
  } else {
    res.status(400).send('invoiceNumber parameter needs to be a number');
  }
})


/**
 * Deletes an invoice by its invoice number.
 * @param {integer} invoiceNumber
 * @returns {string} returns feedback if invalid requests else returns: "invoice deleted succesfully."
 */
app.delete('/delete-invoice/:invoiceNumber', async (req, res) => {
  if (!isNaN(req.params.invoiceNumber)) {
    await pg
      .from('invoices')
      .where({
        invoice_number: req.params.invoiceNumber
      })
      .del()
    res.status(200).send('invoice deleted succesfully.');
  } else {
    res.status(400)
    res.send('invoiceNumber parameter needs to be a number')
  }
})

// create and fill database automatically
async function initialiseTables() {
  await pg.schema.hasTable('invoices').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('invoices', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('reference_number');
          table.string('business_name');
          table.string('client_name');
          table.double('amount_total');
          table.bigInteger('invoice_number').unique();
          table.date('due_date');
          table.integer('type_id');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table invoices');
        });
    }
  });

  await pg.schema.hasTable('invoice_types').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('invoice_types', (table) => {
          table.increments();
          table.string('name');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table invoice_types');
          // fill table with different invoice types
          await pg.table('invoice_types').insert([
            { name: 'Standard invoice' }, 
            { name: 'Credit invoice' },
            { name: 'Expence report' },
            { name: 'Debit invoice' },
            { name: 'Mixed invoice' },
            { name: 'Commercial invoice' },
            { name: 'Timesheet invoice' },
            { name: 'Pro forma invoice' },
            { name: 'Intrim invoice' },
            { name: 'Final invoice' },
            { name: 'Past due invoice' },
            { name: 'Recurring invoice' },
            { name: 'E-invoice' }
          ])
        });
    }
  });
}
initialiseTables();

module.exports = app;